config/v1/types_authentication.go

```diff
@@ -565,27 +601,44 @@ type PrefixedClaimMapping struct {
 	Prefix string `json:"prefix"`
 }
 
+// TokenValidationRuleType defines the type of token validation rule.
+//
+// +kubebuilder:validation:Enum=RequiredClaim;Expression
 type TokenValidationRuleType string
 
 const (
-	TokenValidationRuleTypeRequiredClaim = "RequiredClaim"
+	TokenValidationRuleRequiredClaim TokenValidationRuleType = "RequiredClaim"
+	TokenValidationRuleExpression    TokenValidationRuleType = "Expression"
 )
 
+// TokenClaimValidationRule represents a validation rule based on token claims.
+// If type is RequiredClaim, requiredClaim must be set.
+// If type is Expression, expressionRule must be set.
+//
+// +kubebuilder:validation:XValidation:rule="self.type != 'RequiredClaim' || has(self.requiredClaim)",message="requiredClaim must be set when type is 'RequiredClaim'"
+// +kubebuilder:validation:XValidation:rule="self.type != 'Expression' || has(self.expressionRule)",message="expressionRule must be set when type is 'Expression'"
```

Comment:
I think you are looking for something more like this for these validations: https://github.com/openshift/api/blob/d22fca85e3b9f77f0cf2b43759a5ff927e605f57/example/v1/types_stable.go#L135